lib/base/tlsutility.cpp

```diff
@@ -91,6 +91,10 @@ static void InitSslContext(const Shared<boost::asio::ssl::context>::Ptr& context
 
 	flags |= SSL_OP_CIPHER_SERVER_PREFERENCE;
 
+#ifdef SSL_OP_NO_RENEGOTIATION
+	flags |= SSL_OP_NO_RENEGOTIATION;
+#endif /* SSL_OP_NO_RENEGOTIATION */
```

Comment:
Why the `#ifdef`, where is it active and what happens everywhere else?

Reply:
I'm afraid some distros simply don't have it:

https://github.com/Icinga/icinga2/actions/runs/6625935669/job/18002800324

🤷‍♂️

Reply:
And what happens everywhere else? Is the absence of a `SSL_OP_NO_RENEGOTIATION` macro guaranteed to imply that the OpenSSL version doesn't implement and therefore doesn't offer renegotiation?

An `#ifdef` on the OpenSSL version around the flag in conjunction with a comment why it's safe to not set the flag otherwise would be better. (Check on the version because if there was a change to `SSL_OP_NO_RENEGOTIATION` in the future, that would result in an error to look into instead of the flag silently being omitted.)

Reply:
> Is the absence of a `SSL_OP_NO_RENEGOTIATION` macro guaranteed to imply that the OpenSSL version doesn't implement and therefore doesn't offer renegotiation?

Absolutely not! The option I use here was introduced in v1.1.1:

https://github.com/openssl/openssl/commit/db0f35dda18403accabe98e7780f3dfc516f49de#diff-38e1eecb6013c2fe8bf194e5020ce1afd6e31c371d7849e041464ad822fe745bR1926

But SSL_renegotiate() which has been changed already exists for centuries.

Reply:
So you're saying OpenSSL < 1.1.1 supports renegotiation and there is no way to disable it? And therefore, some old distributions will still happily renegotiate with this PR?

Reply:
Exactly. Oh, almost forgot SLES 12.5 which is not on the list I've linked above:

https://github.com/Icinga/icinga2/actions/runs/6625935626/job/17998216268

Reply:
FYI: https://github.com/postgres/postgres/commit/01e6f1a842f406170e5f717305e4a6cf0e84b3ee (see especially the diff)

Reply:
Also, the problematic CentOS 7 and SLES 12.5 will EOL next year anyway.

Reply:
My test have shown that SSL3_FLAGS_NO_RENEGOTIATE_CIPHERS is not a drop-in replacement.

Reply:
* Test subject: 1db8658942cfd80ad04dbf014567797347b971ef
* Platform: CentOS 7
* Result: renegotiation works

Reply:
> My test have shown that SSL3_FLAGS_NO_RENEGOTIATE_CIPHERS is not a drop-in replacement.

It shouldn't be that.

The issue I linked adds the flag to `connection->s3->flags` and says that the problem with 1.1.0 is that this `flags` member is no longer accessible. So you can't just pass it to `SSL_CTX_set_options()` like you do:

https://github.com/Icinga/icinga2/blob/1db8658942cfd80ad04dbf014567797347b971ef/lib/base/tlsutility.cpp#L94-L100